instrumentation/jdbc/library/src/main/java/io/opentelemetry/instrumentation/jdbc/internal/CheckedCallable.java

```diff
@@ -0,0 +1,12 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.instrumentation.jdbc.internal;
+
+@FunctionalInterface
+public interface CheckedCallable<T, E extends Throwable> {
```

Comment:
Isn't normal Java 8 `Callable` checked (declares Exception in signature)?

Reply:
Actually, it's bad naming. In my case, I need a Callable with a generic error. If I change to the JVM's `Callable` I will need to wrap with additional try/catch in `wrapCall` method or in all other methods.

Example:

```java
  // change to use JVM's Callable instead of CheckedCallable
  private <T> T wrapCall(Callable<T> callable) throws Exception {
    Context parentContext = Context.current();

    if (!Span.fromContext(parentContext).getSpanContext().isValid()) {
      // this instrumentation is already very noisy, and calls to getConnection outside of an
      // existing trace do not tend to be very interesting
      return callable.call();
    }

    Context context = instrumenter().start(parentContext, delegate);
    T result;
    try (Scope ignored = context.makeCurrent()) {
      result = callable.call();
    } catch (Exception t) {
      instrumenter().end(context, delegate, null, t);
      throw t;
    }
    instrumenter().end(context, delegate, null, null);
    return result;
  }
```

As result, I will need to add try/catch to other methods which use `wrapCall`, but this catch will never executed:

<img width="855" alt="CleanShot 2021-06-25 at 15 59 17@2x" src="https://user-images.githubusercontent.com/139017/123391356-84a02880-d5ce-11eb-932d-184b6563cdaa.png">

Another way is to change `wrapCall` method to catch `Exception`:

```java
  private <T> T wrapCall(Callable<T> callable) throws SQLException {
    Context parentContext = Context.current();

    if (!Span.fromContext(parentContext).getSpanContext().isValid()) {
      // this instrumentation is already very noisy, and calls to getConnection outside of an
      // existing trace do not tend to be very interesting
      try {
        return callable.call();
      } catch (SQLException e) {
        throw e;
      } catch (Exception e) {
        // must never happens
        throw new SQLException(e);
      }
    }

    Context context = instrumenter().start(parentContext, delegate);
    T result;
    try (Scope ignored = context.makeCurrent()) {
      result = callable.call();
    } catch (Exception e) {
      instrumenter().end(context, delegate, null, e);
      if (e instanceof SQLException) {
        throw (SQLException) e;
      }
      // must never happens
      throw new SQLException(e);
    }
    instrumenter().end(context, delegate, null, null);
    return result;
  }
```

For me both of these approaches look a little bit ugly, that's why I chose the `CheckedCallable` way.

But I don't mind to change if you think so.

Reply:
Nitpicking: you can rename it to [`ThrowingSupplier`](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/testing-common/src/main/java/io/opentelemetry/instrumentation/testing/util/ThrowingSupplier.java)

Reply:
oh, @mateuszrzeszutek, big thanks for this!

Reply:
Done.